src/main/java/io/anserini/search/SimpleImpactSearcher.java

```diff
@@ -0,0 +1,425 @@
+        this.docid = docid;
+        this.lucene_docid = lucene_docid;
+        this.score = score;
+        this.contents = contents;
+        this.raw = raw;
+        this.lucene_document = lucene_document;
+      }
+    }
+  
+    protected SimpleImpactSearcher() {
+    }
+  
+    /**
+     * Creates a {@code SimpleImpactSearcher}.
+     *
+     * @param indexDir index directory
+     * @throws IOException if errors encountered during initialization
+     */
+    public SimpleImpactSearcher(String indexDir) throws IOException {
+      Path indexPath = Paths.get(indexDir);
+  
+      if (!Files.exists(indexPath) || !Files.isDirectory(indexPath) || !Files.isReadable(indexPath)) {
+        throw new IllegalArgumentException(indexDir + " does not exist or is not a directory.");
+      }
+    
+      this.reader = DirectoryReader.open(FSDirectory.open(indexPath));
+      // Default to using ImpactSimilarity.
+      this.similarity = new ImpactSimilarity();
+      this.generator = new BagOfWordsQueryGenerator();
+      cascade = new RerankerCascade();
+      cascade.add(new ScoreTiesAdjusterReranker());
+    }
+  
+  
+    /**
+     * Returns the number of documents in the index.
+     *
+     * @return the number of documents in the index
+     */
+     public int getTotalNumDocuments(){
+       // Create an IndexSearch only once. Note that the object is thread safe.
+       if (searcher == null) {
+         searcher = new IndexSearcher(reader);
+         searcher.setSimilarity(similarity);
+       }
+  
+       return searcher.getIndexReader().maxDoc();
+     }
+  
+    /**
+     * Closes this searcher.
+     */
+    @Override
+    public void close() throws IOException {
+      try {
+        reader.close();
+      } catch (Exception e) {
+        // Eat any exceptions.
+        return;
+      }
+    }
+  
+    /**
+     * Searches the collection using multiple threads.
+     *
+     * @param queries list of queries
+     * @param qids list of unique query ids
+     * @param k number of hits
+     * @param threads number of threads
+     * @return a map of query id to search results
+     */
+    public Map<String, Result[]> batchSearch(List<Map<String, Float>> queries, List<String> qids, int k, int threads) {
+      return batchSearchFields(queries, qids, k, threads, new HashMap<>());
+    }
+  
+    /**
+     * Searches the provided fields weighted by their boosts, using multiple threads.
+     * Batch version of {@link #searchFields(String, Map, int)}.
+     *
+     * @param queries list of queries
+     * @param qids list of unique query ids
+     * @param k number of hits
+     * @param threads number of threads
+     * @param fields map of fields to search with weights
+     * @return a map of query id to search results
+     */
+    public Map<String, Result[]> batchSearchFields(List<Map<String, Float>> queries, List<String> qids, int k, int threads,
```

Comment:
Let's not implement fielded search yet until we have a realistic use case? Otherwise, it's dead code that's untested...